### PR TITLE
React app

### PR DESCRIPTION
**Pull request by manutuero**
### Summary

I upload a Tic Tac Toe made in React, proposed in this tutorial: [https://reactjs.org/tutorial/tutorial.html](url)

### Trello Card
https://trello.com/c/i330r4xC

### Screenshot
tic-tac-toe

### CODE REVIEW

File: react-app/src/app/components/Board/index.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/* eslint-disable no-use-before-define */
+import React, { Component } from 'react';
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: `import React, { Component, Fragment } from 'react';`

And use Fragment directly

File: react-app/src/app/components/Board/index.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import './styles.css';
+
+class Board extends Component {
+  renderSquare = i => <Square value={this.props.squares[i]} onClick={() => this.props.onClick(i)} />;
+
+  render() {
+    return (
+      <React.Fragment>
+        <div className={status}>{status}</div>
+        <div className={'board-row'}>
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: `className="board-row"`

File: react-app/src/app/components/Board/index.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          {this.renderSquare(0)}
+          {this.renderSquare(1)}
+          {this.renderSquare(2)}
+        </div>
+        <div className={'board-row'}>
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: same

File: react-app/src/app/components/Board/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+          {this.renderSquare(3)}
+          {this.renderSquare(4)}
+          {this.renderSquare(5)}
+        </div>
+        <div className={'board-row'}>
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: same

File: react-app/src/app/components/Board/index.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+import './styles.css';
+
+class Board extends Component {
+  renderSquare = i => <Square value={this.props.squares[i]} onClick={() => this.props.onClick(i)} />;
+
+  render() {
+    return (
+      <React.Fragment>
+        <div className={status}>{status}</div>
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: Where does this status came from?

File: react-app/src/app/components/Square/index.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import React from 'react';
+import PropTypes from 'prop-types';
+
+import './styles.css';
+
+const Square = props => (
+  <button className={'square'} onClick={props.onClick}>
+    {props.value}
+  </button>
+);
```
Comment (LucasZibell) [inline react-app/src/app/components/Square/index.js]: Use functions for components.
```
function Square({ value, onClick }) {
return(
  <button className={'square'} onClick={props.onClick}>
    {props.value}
  </button>
);
}
```

File: react-app/src/app/components/Game/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+          <Board squares={current.squares} onClick={i => this.handleClick(i)} />
+          <div>{status}</div>
+          <ol>{moves}</ol>
+        </div>
+      </div>
+    );
+  }
+}
+
+// Helper function
```
Comment (LucasZibell) [inline react-app/src/app/components/Game/index.js]: Remove comment

File: react-app/src/app/components/Game/index.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  const lines = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
+  // eslint-disable-next-line no-plusplus
+  for (let i = 0; i < lines.length; i++) {
+    const [a, b, c] = lines[i];
+    if (squares[a] && squares[a] === squares[b] && squares[a] === squares[c]) {
+      return squares[a];
+    }
+  }
+  return null;
+}
```
Comment (LucasZibell) [inline react-app/src/app/components/Game/index.js]: move this function to a utils.js file

File: react-app/src/app/components/Game/index.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    });
+
+    return (
+      <div className={'game'}>
+        <div className={'game-info'}>
+          <Board squares={current.squares} onClick={i => this.handleClick(i)} />
+          <div>{status}</div>
+          <ol>{moves}</ol>
+        </div>
+      </div>
```
Comment (LucasZibell) [inline react-app/src/app/components/Game/index.js]: This component is getting pretty large, divide it into index.js to contain the logic and layout.js to contain the html. Then you can import here the layout and render it

File: react-app/src/app/components/Game/index.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    );
+  }
+}
+
+// Helper function
+function calculateWinner(squares) {
+  const lines = [[0, 1, 2], [3, 4, 5], [6, 7, 8], [0, 3, 6], [1, 4, 7], [2, 5, 8], [0, 4, 8], [2, 4, 6]];
+  // eslint-disable-next-line no-plusplus
+  for (let i = 0; i < lines.length; i++) {
```
Comment (LucasZibell) [inline react-app/src/app/components/Game/index.js]: use `forEach`

File: react-app/src/index.js [hunk 11]
```diff
@@ -0,0 +1,10 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import './scss/index.scss';
+import Game from '../src/app/components/Game'; // eslint-disable-line import/first
+
+import registerServiceWorker from './registerServiceWorker';
+
+ReactDOM.render(<Game />, document.getElementById('root'));
```
Comment (LucasZibell) [inline react-app/src/index.js:9]: Import `Game` from a screen folder

**Review by LucasZibell — CHANGES_REQUESTED** [review]

File: react-app/src/app/components/Board/index.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+import Square from '../Square';
+
+import './styles.css';
+
+class Board extends Component {
+  static propTypes = {
+    squares: PropTypes.Array,
+    onClick: PropTypes.func
+  };
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: This should be:
```
Board.propTypes = {
    squares: PropTypes.Array,
    onClick: PropTypes.func
}
```
Outside the class

File: react-app/src/app/components/Board/index.js [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+  renderSquare = i => <Square value={this.props.squares[i]} onClick={() => this.props.onClick(i)} />;
+
+  render() {
+    return (
+      <Fragment>
+        <div className="status">{status}</div>
```
Comment (LucasZibell) [inline react-app/src/app/components/Board/index.js]: status?

File: react-app/src/app/components/Square/index.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+  return (
+    <button className="square" onClick={onClick}>
+      {value}
+    </button>
+  );
+}
+
+Square.propTypes = {
+  value: PropTypes.string,
+  onClick: PropTypes.func
```
Comment (LucasZibell) [inline react-app/src/app/components/Square/index.js]: `PropTypes.func.isRequired`

File: react-app/src/app/components/Board/index.js (hunk 12 above)
Comment (LucasZibell) [reply react-app/src/app/components/Board/index.js]: Also, Array is pretty generic. You should use `arrayOf(PropTypes.String)` For example

File: react-app/src/app/screens/Game/layout.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+import Board from '../../components/Board';
+
+import './styles.css';
+
+export function ListItem({ desc, onClick }) {
+  return (
+    <li>
+      <button onClick={onClick}>{desc}</button>
+    </li>
```
Comment (LucasZibell) [inline react-app/src/app/screens/Game/layout.js]: Why the `<li>` tags??

Comment (LucasZibell) [reply react-app/src/app/screens/Game/layout.js]: Move listItem to the component folder

**Review by LucasZibell — CHANGES_REQUESTED** [review]

File: react-app/src/app/screens/Game/layout.js (hunk 15 above)
Comment (manutuero) [reply react-app/src/app/screens/Game/layout.js]: The `<li>` tags were already in the example.

Comment (manutuero) [reply react-app/src/app/screens/Game/layout.js]: If I remove it, the steps are not seen as a vertical list.

Comment (LucasZibell) [reply react-app/src/app/screens/Game/layout.js]: How about a `<div/>` with column class?

**Review by LucasZibell — APPROVED** [review]

**Pull request merged**